### ISSUE TITLE
SPARQL CONSTRUCT queries do not remove duplicates

### ISSUE THREAD

**Issue by dachafra** [oxigraph/oxigraph#525]
**Describe the bug**
A SPARQL CONSTRUCT query does not remove potential duplicates of the output graph, which is not correct as an RDF graph is a set of RDF triples.

**To Reproduce**
1. Oxigraph v0.3.16 (pyOxigraph module) + Ubuntu server
2. Any query that starts with CONSTRUCT clause and generates duplicates from the input RDF graph

**Comment by dachafra** [comment]
Hi @Tpt,
Thank you very much for the quick answer!

Indeed, we are using CONSTRUCT queries and testing the performance and scalability of different triplestores (research work), and in our opinion, it would not be very fair to compare all of them if they differ in the tasks they are performing.

I've been working on constructing RDF (but from mapping rules: RML, R2RML) and some of the most interesting optimizations we made were towards the de-duplication of the graph, for many reasons, but mostly to reduce the number of unnecessary generated triples and the size of the final graph. 

I cannot share the data we are using now but as an example: we are obtaining almost 100K triples with Oxigraph when the actual RDF set is around 15K triples.

**Comment by Tpt** [comment]
Thank you @dachafra!

+1 for bench fairness. Oxigraph performances are currently pretty bad so I am not sure how much it will change results but it is indeed a relevant point. If you want to can "hack" and write `CONSTRUCT { MY_TRIPLES } WHERE { SELECT DISTINCT ?MY_INTERESTING_VARIABLES { MY_PATTERN } }` instead of `CONSTRUCT { MY_TRIPLES } WHERE { MY_PATTERN }` to do some deduplication.

**Comment by Tpt** [comment]
If it's fine with you I am closing this issue because I believe I won't change Oxigraph behavior. Feel free to ask to open it again if you want so.

**Comment by bastiion** [comment]
this issue should be reopened. In any graph where there are cyclic paths, a construct query stumbling over those cycles, which is perfectly valid in the SPARQL graph world, will get a ton of data, because of the missing deduplication produces endless duplicates. A simple query leads to >32MB of data. Queries that are not hand crafted will make requests, like that all the time. I will investigate by myself where the optimization should happen. it could be an option (either oxigraph server settings or per request).

I could give some example queries.

**Comment by Tpt** [comment]
>In any graph where there are cyclic paths, a construct query stumbling over those cycles, which is perfectly valid in the SPARQL graph world, will get a ton of data, because of the missing deduplication produces endless duplicates.

Yes. That's a good point. But making deduplication the default is not very nice because it means all queries will have to pay its price. A quick workaround is to write `CONSTRUCT { my_construct } WHERE {SELECT DISTINCT * WHERE { my_where }}` instead of `CONSTRUCT { my_construct } WHERE { my_where }` to make sure tuples are duplicated before constructing the triples.

**Comment by marcelotto** [comment]
It's very unfortunate, that this issue is being closed without further pursuit. While I greatly appreciate this project, this particular issue is the main reason I had to move away from using Oxigraph as the reference store for the development of [Ontogen](https://ontogen.io), a new version control system for RDF that I recently released. Moreover, I feel compelled to caution Ontogen users against versioning their datasets in Oxigraph due to this problem.

To illustrate the severity of the issue: a CONSTRUCT query returning approximately 150 triples for the history of a small, synthetic test dataset becomes bloated with duplicates, resulting in an ~8MB large result. This renders it impractical for use of Ontogen with realistic dataset sizes.

**Comment by Tpt** [comment]
@marcelotto thank you for commenting. Ontogen seems like a great project! If I understand correctly, you mean that Jena returns less triples than Oxigraph? Does Jena deduplicate blank nodes or is it only avoiding to emit duplicated triples? For the second case Oxigraph can and probably should try to be smarter and do some deduplication

**Comment by Tpt** [comment]
Thank you! Removing duplicated blank node would change the RDF graph so I guess Jena is not doing that.

**Comment by marcelotto** [comment]
Just remembered, that I did some profiling, that it isn't the query performance taking so long (because I first thought I'd have to optimize the query that took so long). The results showed that the deserializing of these large results took most of the time.

**Comment by bastiion** [comment]
We are facing a similar problem, for a software that extracts documents from a graph in a user configurable depth - that means all relations are beeing fetched deeply - thus leading to a high chance of duplicate triples. I do understand, that when using a streaming approach where you want to get results even if the query has not finished yet, it is not feasible to de-depuplicate - but what's the problem on providing an option - maybe **within the Header of  the HTTP-Request** ,where you can control, whether you wan't de duplication or not.  In fact, the default should be de-duplicated, if a flag for speed, or streaming is set it can disable de-duplication.

**Comment by Tpt** [comment]
My current idea is to create a cache of the last emitted triples that would be used for streaming partial deduplication (triples are emitted on the fly except if they are already in the cache). This way I hope most duplicated triples will be removed while having only a bounded memory overhead.

**Comment by marcelotto** [comment]
I can confirm that this issue has been resolved with the release of Oxigraph v0.4. I have tested with the previously problematic queries in Ontogen and no longer get any duplicates. 👏

Thank you very much. 🙏🏻

**Comment by Tpt** [comment]
Amazing! Note that if the output is too big some duplicates might still be emitted.
